File: sdks/js/src/compat.ts
/** Unstructured API compatibility — drop-in replacement for unstructured-client. */

import type { Element } from "./types.js";
import { DocParseError } from "./types.js";

const DEFAULT_BASE_URL = "https://ailang-dev-docparse-api-ejjw6zt3bq-ew.a.run.app";

class GeneralApi {
  private baseUrl: string;
  private apiKey: string;
  private timeout: number;

  constructor(baseUrl: string, apiKey: string, timeout: number) {
    this.baseUrl = baseUrl;
    this.apiKey = apiKey;
    this.timeout = timeout;
  }

  /** Partition a document — returns Unstructured-format elements. */
  async partition(opts: { file: string; strategy?: string }): Promise<Element[]> {
    const url = `${this.baseUrl}/general/v0/general`;
    const headers: Record<string, string> = {
      "Content-Type": "application/json",
    };
    if (this.apiKey) {
      headers["unstructured-api-key"] = this.apiKey;
    }

    const resp = await fetch(url, {
      method: "POST",
      headers,
      body: JSON.stringify({ args: [opts.file, opts.strategy || "auto"] }),
    });

    if (!resp.ok) throw new DocParseError(`API error: ${resp.status}`, resp.status);

    const outer = await resp.json();
    if (outer.error) throw new DocParseError(outer.error);

    const resultStr = outer.result || "[]";
    try {
      const elements = JSON.parse(resultStr);
      return Array.isArray(elements) ? elements : [];
    } catch {
      return [];
    }
  }
}

/**
 * Drop-in replacement for `unstructured-client`'s UnstructuredClient.
 *
 * Migration:
 * ```typescript
 * // Before
 * import { UnstructuredClient } from 'unstructured-client';
 * const client = new UnstructuredClient({ serverUrl: 'https://api.unstructured.io' });
 *
 * // After — one import change
 * import { UnstructuredClient } from '@sunholo/docparse/compat';
 * const client = new UnstructuredClient({
 *   serverUrl: 'https://ailang-dev-docparse-api-ejjw6zt3bq-ew.a.run.app'
 * });
 * ```
 */
export class UnstructuredClient {
  general: GeneralApi;

  constructor(opts: { serverUrl?: string; apiKey?: string; timeout?: number }) {
    const baseUrl = (opts.serverUrl || DEFAULT_BASE_URL).replace(/\/$/, "");
    this.general = new GeneralApi(baseUrl, opts.apiKey || "", opts.timeout || 60000);
  }
}

File: sdks/go/compat.go
package docparse

import (
	"context"
	"encoding/json"
	"fmt"
)

// UnstructuredClient is a drop-in replacement for the Unstructured API client.
//
// Migration:
//
//	// Before
//	client := unstructured.New("https://api.unstructured.io")
//
//	// After
//	client := docparse.NewUnstructuredClient("https://ailang-dev-docparse-api-ejjw6zt3bq-ew.a.run.app")
type UnstructuredClient struct {
	client *Client
}

// NewUnstructuredClient creates an Unstructured-compatible client.
func NewUnstructuredClient(serverURL string, apiKey ...string) *UnstructuredClient {
	key := ""
	if len(apiKey) > 0 {
		key = apiKey[0]
	}
	return &UnstructuredClient{
		client: New(key, WithBaseURL(serverURL)),
	}
}

// Partition partitions a document into Unstructured-format elements.
func (uc *UnstructuredClient) Partition(ctx context.Context, filepath string, strategy ...string) ([]Element, error) {
	strat := "auto"
	if len(strategy) > 0 {
		strat = strategy[0]
	}

	data, err := uc.client.call(ctx, "POST", "/general/v0/general", []string{filepath, strat})
	if err != nil {
		return nil, err
	}

	var elements []Element
	if err := json.Unmarshal(data, &elements); err != nil {
		return nil, fmt.Errorf("unmarshal elements: %w", err)
	}
	return elements, nil
}
